module/zcommon/zfs_fletcher_aarch64_neon_impl.S

```diff
@@ -0,0 +1,113 @@
+/*
+ * Optimize the neon implementation which is generated by GCC12
+ * or Clang by the commit: https://github.com/openzfs/zfs/pull/14234
+
+ * Due to the compiler limitation, the GCC version before 12 will
+ * produce not optimized code for Aarch64, so use the asm function
+ * to accelerate the performance here.
+ * Copyright (C) 2022 Kevin Zhao.
+ *
+ * Authors:
+ *	Kevin Zhao <kevin.zhao@linaro.org>
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * OpenIB.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *	- Redistributions of source code must retain the above
+ *	  copyright notice, this list of conditions and the following
+ *	  disclaimer.
+ *
+ *	- Redistributions in binary form must reproduce the above
+ *	  copyright notice, this list of conditions and the following
+ *	  disclaimer in the documentation and/or other materials
+ *	  provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if defined(__aarch64__)
+	.section .text
+	.arch armv8-a
+	.globl	fletcher_4_aarch64_neon_native_impl
+	.type	fletcher_4_aarch64_neon_native_impl, @function
+fletcher_4_aarch64_neon_native_impl:
+		.cfi_startproc
+        ldp     q4, q0, [x0]
+        cmp     x2, #1
+        ldp     q5, q1, [x0, #32]
+        ldp     q6, q2, [x0, #64]
+        ldp     q7, q3, [x0, #96]
+        b.lt    .LBB0_3
```

Comment:
The cmp and b.lt instructions can be dropped because the first loop iteration will always execute. The .LBB0_3 label can be removed too.

The same can be done with the byteswap version.

Reply:
The goldbot link in my comment from 5 days ago had this optimization, so I am a little confused why it is missing from here:

https://github.com/openzfs/zfs/pull/14219#issuecomment-1334427948